Parse/src/main/java/com/parse/ParseLogInterceptor.java

```diff
@@ -93,67 +109,60 @@ public void writeLine(String str) {
   }
 
   private static class ProxyInputStream extends InputStream {
+    // Helper stream to proxy the original input stream to other input stream
+    InputStream originalInput;
+    PipedInputStream proxyInput;
+    PipedOutputStream proxyOutput;
+
+    public ProxyInputStream(
+        InputStream originalInput, final InterceptCallback callback) throws IOException {
+      this.originalInput = originalInput;
+      proxyInput = new PipedInputStream();
+      try {
+        proxyOutput = new PipedOutputStream(proxyInput);
+      } catch (IOException e) {
+        callback.done(null, e);
+        ParseIOUtils.closeQuietly(proxyOutput);
+        ParseIOUtils.closeQuietly(proxyInput);
+        throw e;
+      }
 
-    private ParseHttpResponse response;
-    private String requestId;
-    private boolean hasBeenPrinted;
-    private ByteArrayOutputStream bodyOutput;
-    private Logger logger;
-    private boolean needsToBePrinted;
-
-    public ProxyInputStream(String requestId, ParseHttpResponse response, Logger logger)
-        throws FileNotFoundException {
-      this.requestId = requestId;
-      this.response = response;
-      this.logger = logger;
-      bodyOutput = new ByteArrayOutputStream();
-      needsToBePrinted = isContentTypePrintable(response.getContentType());
+      // We need to make sure we read and write proxyInput/Output in separate thread, otherwise
+      // there will be deadlock.
+      Task.call(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          callback.done(proxyInput, null);
+          return null;
+        }
+      }, ParseExecutors.io());
```

Comment:
Is there a more elegant way to do this? A Callback and background thread seems super hacky.